Updating instructions in factors README and comments in starter template (#564)
risc0/risc0: examples/factors/README.md

cargo risczero new factors
```
This will create a project named `factors` in the directory where you ran the `cargo risczero new` command. Now we can enter our new project's directory and start working on it!
```sh
cd factors
```

## Step 2: Give the methods package a name

The methods package contains the program that executes on the guest zkVM as well as a few supporting libraries. Before we proceed, let's change its name in `methods/Cargo.toml` to match the name of our project, renaming `name = methods` to `name = factors-methods`.

Parts of this package are included in the driver program `src/main.rs`, so change the line

```no_compile
use methods::{METHOD_NAME_ELF, METHOD_NAME_ID};
```

> 

Reviewer: feels like we're missing a "to" here, but it's hard to parse in this diff format

> 

Reviewer 2: I checked in View File and there does indeed need to be a "to" here.
```suggestion

to

```

Reviewer 2: I think this may have been accidentally resolved, this is not actually in yet.

While we're at it, let's update the references to `METHOD_NAME_ELF` and `METHOD_NAME_ID` in `factors/host/src/main.rs` to reflect our new program name. Don't worry about why these lines are included yet; for now, we're just being diligent not to leave dead references behind.

Here are what the other two lines with `METHOD_NAME_ELF` and `METHOD_NAME_ID` should look like after updating:

```no_compile
    // Rename METHOD_NAME_ELF
    let mut exec = Executor::from_elf(env, MULTIPLY_ELF).unwrap();

    // Rename METHOD_NAME_ID
    // receipt.verify(MULTIPLY_ID).unwrap();
```

### Step 4: Build and run the project!

In the main project folder, build and run the project using `cargo run --release`.
Nothing exciting will happen yet, but it should give you an error if you made one of the above changes incorrectly.
Use this command any time you'd like to check your progress.

## Concept break: How do we run and prove the guest program?

Our next objective is to provide the guest program with input. Before we implement this, let's take a closer look at how we run and prove the guest program in `factors/src/main.rs`.

> In the starter template project, our host driver program creates an executor environment before constructing an executor.  When the `Executor::run()` command executes, it will produce a session. The `Session::prove()` command will cause our guest program to execute, producing a receipt:

Reviewer: ```suggestion
In the starter template project, our host driver program creates an executor environment before constructing an executor.  When `Executor::run()` is called, it will produce a session. Calling `Session::prove()` will cause our guest program to execute, producing a receipt:
```

Here are what the other two lines with `METHOD_NAME_ELF` and `METHOD_NAME_ID` should look like after updating:

```rust
fn main() {
    // Here we make an executor, loading the (renamed) ELF binary.

>     let mut exec = Executor::from_elf(env, METHOD_NAME_ELF).unwrap();

Reviewer: This takes an `env` parameter, which isn't defined anywhere.

Here's a suggestion, let's figure out how to make this README something that gets run as part of doc builds, which would then give us the ability to 'unit test' this example code.

Author: Looking into this and working out errors right now -- before I continue, is this an improvement we should be aiming to include in today's release?

Reviewer: I think we should try to get as much of our docs under test as possible

Reviewer 2: Looking at this thread: https://github.com/rust-lang/cargo/issues/383, it seems like built in Rust tooling won't accomplish this. It includes a comment suggesting a fairly straightforward macro for adding README tests: https://github.com/rust-lang/cargo/issues/383#issuecomment-720873790, however I haven't gotten this working -- I'll take a less rushed look tomorrow.

Author: That's the one I've been trying out and running into errors, going to keep troubleshooting today.

Author: More detailed version: `#![doc = include_str!("../README.md")]` is included in most of our crates under `src/lib.rs`, but I'm getting fingerprint errors that seem to be complaining about files changing out from under the compiler when I add the line to the factors example under `src/main.rs`. I'm going to pull my branch down and rerun in case I've done something funny locally, but it's not yet complaining about the syntax of various README snippets, my marker for at least having the README successfully included.

Author: The docs are building now so the branch issue was probably just git flow cruft but it's strange that only the digital signature example README is sensitive to asserting untruths.

Author: From convo with @flaub for now will add factors function and tests into a `lib.rs` to "solve" this problem but will come back to this (and see what else is out there with untested README snippets).

Author: Working through decisions around how to present code snippets to get this functional but still educational. The one that's most displeasing is aligning with our per-example naming conventions. We change `methods` to `factors-methods`, and thus
`use methods::{MULTIPLY_ID, MULTIPLY_ELF};` fails as an import unless I tell people to first complete this rename and get `use factor_methods::{MULTIPLY_ID, MULTIPLY_ELF};`.

I love that we've cut a clean path in our README to "minimal steps necessary to successfully build" to get people into the edit-compile loop as quickly as possible. Injecting this rename task breaks that a tiny bit, but I'm writing out instructions to feel it out.

Author: A second sub-ideal situation is that the final product is now split into `main.rs` and `lib.rs`, so there's not a pleasant 1:1 to follow along with. That said, the cost of having outdated README here is sufficiently high to warrant proceeding. There might be room to tack on a "let's make a library" step. It just won't give folks an instant visual clue. 

Loving @tzerrell 's suggestion to just include the README and nothing else. Tradeoff here is not teaching "let's make a library" so I'll leave it off for now but it's an option to consider if we strongly dislike the complexity of this next revision.

Here are what the other two lines with `METHOD_NAME_ELF` and `METHOD_NAME_ID` should look like after updating:

```rust
    let mut env =

>         ExecutorEnv::new(MULTIPLY_ELF).expect("Prover should be constructed from valid ELF binary");

Reviewer 2: ```suggestion
        ExecutorEnv::new(MULTIPLY_ELF).expect("ExecutorEnv should be constructed from valid ELF binary");
```

Reviewer 2: Thanks! I clearly didn't really turn my brain on for this part of the review.

So yeah, suggestion is instead use the code that actually works from the factors example and describe it here.

Author: I can make the correct update and check the rest of the factors README for strays as well (MULTIPLY_ELF along with env as params for Executor construction).

    let b: u64 = 23;
```

Currently, our host driver program creates an executor environment before constructing an executor. The `Executor::run()` command will produce a session. The `Session::prove()` command will cause our guest program to execute, producing a receipt:

```rust
    // First, we construct an executor environment
    let env = ExecutorEnv::default();

    // TODO: add guest input to the executor environment using

>     // ExecutorEnv::add_input()

Reviewer 2: If this is also wrong in the in-code comment, that will need to be updated as well.

```suggestion
    // ExecutorEnvBuilder::add_input()
```

    // Next, we make an executor, loading the (renamed) ELF binary.
    let mut exec = Executor::from_elf(env, METHOD_NAME_ELF).unwrap();

    // Run the executor to produce a session.
    let session = exec.run().unwrap();

    // Prove the session to produce a receipt.
    let receipt = session.prove().unwrap();
```
 We'd like the host to make the values of `a` and `b` available to the guest prior to execution.
 Because the executor environment is responsible for managing guest-readable memory, we need to add these values as inputs after the environment is created. We do this in between making the executor environment and constructing the executor (in place of the "`TODO`" in the previous code snippet).

 As a preliminary step, let's first move away from the default construction `ExecutorEnv::default()` and instead use `ExecutorEnv::builder()`, which supports adding guest input:

 ```rust
    // First, we construct an executor environment
      let env = ExecutorEnv::builder().build();

          // TODO: add guest input to the executor environment using

>     // ExecutorEnv::add_input()

Reviewer 2: ```suggestion
    // ExecutorEnvBuilder::add_input()
```

Here is the complete guest program.
We'll break this down step by step below:

> ```no_compile

Author: This hasn't worked for me but I will change if I find a workaround that allows them both.

Now you should be able to confirm your work again with `cargo run --release`.
Once more, the program won't do anything, but it should run successfully and build with a single warning about `from_slice` being unused.

## Step 7 (Host): Generate a receipt and read its journal contents

For this step, we return to the main file for the host driver program at `factors/host/src/main.rs`, which currently has a placeholder comment asking to fill in with code for handling the [receipt]:

```no_compile
    // Run the executor to produce a session.
    let session = exec.run().unwrap();

    // Prove the session to produce a receipt.
    let receipt = session.prove().unwrap();

    // TODO: Implement code for transmitting or serializing the receipt for
    // other parties to verify here

> 

Author: This isn't in the `factors` example but it's present in lines 31-32 of `templates/rust-starter/host/src/main.rs`.I'll grant you though it's an odd step to instruct on now that in the `factors` example the journal content extraction is in `multiply_factors` and the journal verification is in `main.rs`. Still for now I think since it lines up reasonably well with the original template content, it's better to leave it.

```

In a real-world scenario, we'd want to hand the [receipt] to someone else, but reading it ourselves will be a nice way to check our project is working as expected.
So, let's extract the [journal]'s contents by replacing the "`TODO`" in the above code snippet with the following lines.

```rust
    use factors_methods::{MULTIPLY_ELF, MULTIPLY_ID};
    use risc0_zkvm::{
      serde::{from_slice, to_vec},
      Executor, ExecutorEnv,
    };

    let a: u64 = 17;
    let b: u64 = 23;

>     let env = ExecutorEnv::builder()
>     // Send a & b to the guest
>     .add_input(&to_vec(&a).unwrap())
>     .add_input(&to_vec(&b).unwrap())
>     .build();

Reviewer: ```suggestion
    // Send a & b to the guest
    let env = ExecutorEnv::builder()
        .add_input(&to_vec(&a).unwrap())
        .add_input(&to_vec(&b).unwrap())
        .build();
```
